### PR TITLE
[SPARK-25447][SQL] Support JSON options by schema_of_json()

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

In the PR, I propose to extended the `schema_of_json()` function, and accept JSON options since they can impact on schema inferring. Purpose is to support the same options that `from_json` can use during schema inferring.

## How was this patch tested?

Added SQL, Python and Scala tests (`JsonExpressionsSuite` and `JsonFunctionsSuite`) that checks JSON options are used.

### CODE REVIEW

**Comment by MaxGekk** [comment]
jenkins, retest this, please

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -3611,6 +3611,20 @@ object functions {
    */
   def schema_of_json(e: Column): Column = withExpr(new SchemaOfJson(e.expr))
 
+  /**
+   * Parses a column containing a JSON string and infers its schema using options.
+   *
+   * @param e a string column containing JSON data.
+   * @param options JSON datasource options that control JSON parsing and type inference.
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: maybe you can say refer to DataFrameReader.json for the list of options

Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Sure but not all options can impact on schema inferring. @rxin Maybe it makes sense to list the options which I pointed out in the ticket: https://issues.apache.org/jira/browse/SPARK-25447 ?

[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: But people wouldn't probably take a look for a ticket. Duplicated documentations are not a good idea as well.

Silently ignoring provided options is worse I guess - we should probably throw an exception for `from_json` too in Spark 3.0.

I thought we are going to do the similar stuff for parse mode as well - `DROPMALFORMED` one throws an exception?

[the remark this answers is not in the document]
Comment (rxin) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: do we fail if users currently specify an option in dataframereader that doesn't apply? if we don't i wouldn't fail here.

[the remark this answers is not in the document]
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: As I see we don't fail . Simple example is if `multiLine` is enabled, `lineSep` is ignored. There are another examples.

**Comment by MaxGekk** [comment]
@rxin @HyukjinKwon @gatorsmile Can we move forward with the PR without throwing an exception for ignored options by `schema_of_json()`?

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -2328,11 +2328,14 @@ def to_json(col, options={}):
...
 @since(2.4)
-def schema_of_json(col):
+def schema_of_json(col, options={}):
     """
     Parses a column containing a JSON string and infers its schema in DDL format.
 
     :param col: string column in json format
+    :param options: options to control parsing. accepts the same options as the JSON datasource
+
+    .. note:: Since Spark 2.5, it accepts options to control schema inferring.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Let's convert this to:

```
.. versionchanged:: 2.5
   it accepts `options` parameter to control schema inferring.
```

**Review by HyukjinKwon — APPROVED** [review]

**Comment by MaxGekk** [comment]
@rxin @gatorsmile Do you have any objections of merging the PR?

**Review by HyukjinKwon — APPROVED** [review]
LGTM

**Comment by HyukjinKwon** [comment]
@rxin and @gatorsmile, I am getting this in but please feel free to let me know if you guys think it shouldn't be added.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by MaxGekk** [comment]
Thank you @HyukjinKwon

**Pull request closed without merging**